### DISCUSSION TITLE
Ethernet ESP32 board support

### DISCUSSION THREAD

**Discussion by schaffman5** [HomeSpan/HomeSpan#49] [Q&A] [answered]
Are there any examples with ethernet connection support instead of wifi?

**Comment by HomeSpan** [comment]
Sorry, HomeSpan has not been configured for use with Ethernet, just WiFi.   This is a good suggestion however and I'll put it on the list for future expansion.  Can I ask what Ethernet board you are using?

**Reply by schaffman5** [reply]
Bummer, thanks for the reply.  I'm using a [Silicognition wESP32](https://wesp32.com) board, which is a very nice PoE board and looks to be perfect for wired Homekit applications.  Creating the ethernet connection and obtaining an IP address is [very straightforward in Arduino](https://wesp32.com/software/#arduino-esp32).  Would handing off to Homespan after that be a major overhaul of the code?

**Comment by HomeSpan** [comment, marked as answer]
I would be a little more complicated than that since each Controller connects as a WiFiClient.  The code would need to be modified so that they each connected as an EthernetClient instead.  Would also need some logic to either detect the Ethernet and use that as the default if connected, or have the user configure with something like homeSpan.useEthernet() at the top of the sketch.   I'll see about acquiring an ESP32 Ethernet board and give it a try.

**Comment by edpgcooper** [comment]
I'd love to see Ethernet support, I'm using WT32-ETH01 boards

**Comment by HomeSpan** [comment]
Agree it would be great to extend HomeSpan to support Ethernet.  I purchased an Ethernet board from Adafruit last month and spent a good amount of time trying to integrate.  One known issue is a point of incompatibility between the latest Arduino Ethernet Server library and the ESP-32 implementation of that library. There are workarounds that require overriding some of the library but I was able to address this directly in HomeSpan in a fairly robust way that solved the problem.   

Unfortunately, once I got the Ethernet reliably working, I uncovered another lesser-known issue:  MDNS problems.   Though the ESP-32 seamlessly supports MDSN (Bonjour) through WiFi UTP, this does not work for Ethernet connectivity.  I tried a number of external libraries but could not find a solution.  It appears to be a core problem in the UTP Ethernet implementation.

HomeKit heavily relies on MDNS broadcasting, so the lack of MDNS support for Ethernet on ESP-32 is a bit of a show-stopper.  If anyone has found a solution to this, please let me know.

**Reply by Xorlent** [reply]
Gregg, are you seeing an issue receiving or sending mDNS from the micro using Ethernet?  One thing the comes to mind is perhaps you can instead specify the broadcast address of 224.0.0.251.  All assuming that you're not already doing this!

**Reply by HomeSpan** [reply]
The issue is with sending (receiving is not needed/used in HomeSpan).  The code seems to work fine, but nothing is ever broadcast.  Based on web research others have had this same issue - it appears to be related to the UDP functionality under Ethernet. However, there have been some significant updates to MDNS in the Arduino-ESP32 version 2.0.2 that seems to have resolved some long-standing MDNS problems in WiFi (which caused a lot of the "not-responding" issues with the Home App).  I will give it another try with my Ethernet board to see if the changes in 2.0.2 address the Ethernet issue as well.
